torchgeo/datasets/ssl4eo_benchmark_landsat.py

```diff
@@ -0,0 +1,348 @@
+
+        self.sample_collection = self.retrieve_sample_collection()
+
+        # train, val, test split
+        np.random.seed(0)
+        sizes = (np.array(self.split_percentages) * len(self.sample_collection)).astype(
+            int
+        )
+        cutoffs = np.cumsum(sizes)[:-1]
+        sample_indices = np.arange(len(self.sample_collection))
+        np.random.shuffle(sample_indices)
+        groups = np.split(sample_indices, cutoffs)
+        split_indices = {"train": groups[0], "val": groups[1], "test": groups[2]}[
+            self.split
+        ]
+
+        self.sample_collection = [self.sample_collection[idx] for idx in split_indices]
+
+    def _verify(self) -> None:
+        """Verify the integrity of the dataset.
+
+        Raises:
+            RuntimeError: if ``download=False`` but dataset is missing or checksum fails
+        """
+        # Check if the extracted files already exist
+        img_pathname = os.path.join(self.root, self.img_dir_name, "**", "all_bands.tif")
+        exists = []
+        for fname in glob.iglob(img_pathname, recursive=True):
+            exists.append(True)
+            break
+        mask_pathname = os.path.join(
+            self.root,
+            self.mask_dir_name,
+            "**",
+            f"{self.mask_product}_{self.year_dict[self.input_sensor]}.tif",
+        )
+        for fname in glob.iglob(mask_pathname, recursive=True):
+            exists.append(True)
+            break
+        if exists and all(exists):
+            return
+
+        # Check if the tar.gz files have already been downloaded
+        exists = []
+        img_pathname = os.path.join(self.root, f"{self.img_dir_name}.tar.gz")
+        if os.path.exists(img_pathname):
+            exists.append(True)
+        mask_pathname = os.path.join(self.root, f"{self.mask_dir_name}.tar.gz")
+        if os.path.exists(mask_pathname):
+            exists.append(True)
+        if exists and all(exists):
+            self._extract()
+            return
+
+        # Check if the user requested to download the dataset
+        if not self.download:
+            raise RuntimeError(
+                f"Dataset not found in `root={self.root}` and `download=False`, "
+                "either specify a different `root` directory or use `download=True` "
+                "to automatically download the dataset."
+            )
+
+        # Download the dataset
+        self._download()
+        self._extract()
+
+    def _download(self) -> None:
+        """Download the dataset."""
+        # download imagery
+        download_url(
+            self.url.format(self.img_dir_name, self.img_dir_name),
+            self.root,
+            md5=self.img_md5s[self.input_sensor] if self.checksum else None,
+        )
+        # download mask
+        download_url(
+            self.url.format(self.mask_dir_name, self.mask_dir_name),
+            self.root,
+            md5=self.mask_md5s[self.input_sensor.split("_")[0]][self.mask_product]
```

Comment:
Not sure why mypy is failing here. I would try calculating some of these variables first instead of doing everything on one line.